pkg/bloomcompactor/bloomcompactor.go

```diff
@@ -354,46 +353,53 @@ func (c *Compactor) compactTenant(ctx context.Context, logger log.Logger, sc sto
 	NGramSkip := c.limits.BloomNGramSkip(tenant)
 	bt, _ := v1.NewBloomTokenizer(c.reg, NGramLength, NGramSkip)
 
-	// TODO: Use ForEachConcurrent?
 	errs := multierror.New()
 	if err := sc.indexShipper.ForEach(ctx, tableName, tenant, func(isMultiTenantIndex bool, idx shipperindex.Index) error {
-		if isMultiTenantIndex {
+		if isMultiTenantIndex { // TODO: handle multitenant tables
 			return fmt.Errorf("unexpected multi-tenant")
 		}
 
+		var indices []Index
 		// TODO: Make these casts safely
 		if err := idx.(*tsdb.TSDBFile).Index.(*tsdb.TSDBIndex).ForSeries(
 			ctx, nil,
 			0, math.MaxInt64, // TODO: Replace with MaxLookBackPeriod
 			func(labels labels.Labels, fingerprint model.Fingerprint, chksMetas []tsdbindex.ChunkMeta) {
-				job := NewJob(tenant, tableName, idx.Path(), fingerprint, labels, chksMetas)
-				jobLogger := log.With(logger, "job", job.String())
+				ownsFingerprint, err := c.sharding.OwnsFp(tenant, uint64(fingerprint))
```

Comment:
Agree, but also my bad: this inefficiency comes from my implementation for `OwnsJob`. I think once we have the tenant, we can create a new`util_ring.NewFingerprintShuffleSharding` object passing the tenant subring to it and we can get rid of all the logic inside sharding.go (I'd move the test inside sharding_test.go to bloomcompactor_test.go tho)

I see one problem with this tho. If we don't refresh the tenant subring, the tenant compaction takes long enough, and there is an ownership change while compacting the tenant, we may end up processing fingerprints for a tenant the instance no longer owns. This may be an over-optimization at this point, but I wonder if we should have some tenant ring timeout that forces us to reload the tenant ring.

Reply:
I would expect that the ring could give us the range of the fingerprints that we own, and after this inside `.ForSeries()` we could just check that FP is within the range, something like: `min_fp <= fingerprint <= max_fp`. without the need to work with the ring inside the callback.

Reply:
> we may end up processing fingerprints for a tenant the instance no longer owns. 
this was discussed during the first weeks and it was agreed that much duplicate work should be fine for a starting point. In the next compaction (eg. in 15 mins) that should be cleared up.  

@vlad-diachenko AFAICT the ring can't give a range, at least as it is. But in the same spirit, we can ask if the job's entire fingerprint range is within the ring by checking the bounds by the ring and go from there. wdyt? 

Overall the goal of this PR is to merge series into a single bloomblock, are we ok to address it in a follow up?

Reply:
@poyzannur I faced a somewhat similar problem when I implemented the bloom gateway client: There I also need to find the fingerprint ranges for each server instance in the ring. 

... while looking at my PR #11336 I saw you commented on that part of the code.